sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,895 @@
+    #[inline(always)]
+    pub unsafe fn get_entry_count_unchecked(data: &[u8]) -> usize {
+        ptr::read_unaligned(data.as_ptr() as *const u64).to_le() as usize
+    }
+
+    /// Returns the number of `SlotHashEntry` items accessible.
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if there are no entries.
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Gets a reference to the `SlotHashEntry` at the specified index.
+    /// Returns `None` if the index is out of bounds.
+    #[inline(always)]
+    pub fn get_entry(&self, index: usize) -> Option<&SlotHashEntry> {
+        self.as_entries_slice().get(index)
+    }
+
+    /// Gets a reference without bounds checking.
+    ///
+    /// # Safety
+    /// Caller must ensure `index < self.len()`.
+    #[inline(always)]
+    pub unsafe fn get_entry_unchecked(&self, index: usize) -> &SlotHashEntry {
+        debug_assert!(index < self.len);
+        let offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+        &*(self.data.deref().as_ptr().add(offset) as *const SlotHashEntry)
+    }
+
+    /// Finds the hash for a specific slot using binary search.
+    ///
+    /// Returns the hash if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    #[inline(always)]
+    pub fn get_hash(&self, target_slot: Slot) -> Option<&[u8; HASH_BYTES]> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot.cmp(&target_slot).reverse())
+            .ok()
+            .map(|index| &entries[index].hash)
+    }
+
+    /// Finds the position (index) of a specific slot using binary search.
+    ///
+    /// Returns the index if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    #[inline(always)]
+    pub fn position(&self, target_slot: Slot) -> Option<usize> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot.cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (either the safe path that called `parse_and_validate_data` or
+    /// the unsafe `new_unchecked`) is responsible for ensuring the slice is big enough
+    /// and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        if self.len == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.data.deref().len() >= NUM_ENTRIES_SIZE + self.len * ENTRY_SIZE);
+
+        let entries_ptr =
+            unsafe { self.data.deref().as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry };
+        unsafe { core::slice::from_raw_parts(entries_ptr, self.len) }
+    }
+}
+
+// Implementation block specific to the safe Ref version
```

Comment:
If we make it genereic over `Deref<Target = [u8]>` we should be able to have a "single" implementation.

Reply:
Addressed in dce5e60da08b583d82c3eee70c6114ba52a2bc8b
and 82336188de4e446676dac8b8cd078bd24f2aa646